### DISCUSSION TITLE
How to obtain a reference to widgets in SimpleComponent / Component?

### DISCUSSION THREAD

**Discussion by gtamas** [Relm4/Relm4#517] [General]
Hi

I'm building an app using the SimpleComponent / Component macro. It works great. But I don't understand how can I refer to the widgets generated by view! macro later?

My understanding is that the update_view() method is automatically handled by the Component, so I can't really use it.
I know I have access to the generated widgets in init() fn using:

let widgets = view_output!();

But what if I need those widgets later, for example in the event handlers in update() fn or in the post_view() fn or elsewhere?

I'm new to Relm4 and Rust, so probably there is some simple solution, but I can't see how this is supposed to work.

**Comment by AaronErhardt** [comment]
If you name widgets with `#[name(my_widget)]` you should be able to access them in `post_view()` or `pre_view()`. Accessing widgets in `update()` is not supported because this is usually a bad pattern, but you can overwrite [`update_with_view`](https://docs.rs/relm4/latest/relm4/component/trait.Component.html#method.update_with_view) if you really need widgets and model at the same time.

**Reply by gtamas** [reply]
Thanks for the reply.

So the problem is I cannot use the` #[name]` or stuff like that, because my UI is dynamically generated. I just have a Frame widget in the view! macro call and its child is generated by a method, which returns a Box containing the form. Something like this:

`view! {
        #[root]
        gtk::Box {
          append: form = &gtk::Frame {
            set_hexpand: true,
            #[track = "model.changed(SomModel::json())"]
            set_child: Some(&model.build_form(&model.json).unwrap())
          }
        }
    }`

That's why I think I need access to the  widgets later. I need to extract the user input from those widgets on form submit. 

So in such a case the update_with_view() is the right way to go? I also think that I cannot use SimpleComponent in this case, but I have to implement Component. Is that correct?

Again, thanks in advance for any help. I'm totally new to Rust and GTK4 and Relm4 so probably there is a much better way to implement all this. Please let me know!

**Reply by AaronErhardt** [reply]
You could do the `#[local]` attribute to use a local variable from the `init` method. This will work similarly to `#[name(...)]` and "my_box" in the following example should be available as field of the generated `Widgets` type.

```rust
view! {
    gtk::Box {
        ... gtk::Frame {
            #[local]
            my_box -> gtk::Box {},
        }
    }
}

fn init(...) {
    let my_box = model.build_form(...);
    let widgets = view_output!();
    // ...
}
```

This means that you can also use `pre_view()` and `post_view()` to access the widgets. `update_with_view()` in the `Component` trait is also an option, but it is usually not necessary to mix state and widgets.

You can also add your own custom fields to the generated `Widgets` type btw.: https://relm4.org/book/stable/component_macro/reference.html#add-more-fields-to-your-widgets

> I need to extract the user input from those widgets on form submit.

Since 0.6, Relm4 has bindings which might be even better for your use-case: https://docs.rs/relm4/latest/relm4/binding/index.html

Basically, the idea is, that if UI state and application state are identical (which is usually the case in forms) you can just have a binding in your model that is synchronized with a property of the UI (such as the state of a toggle button). Something similar is already possible for `gtk::Entry` by storing an `EntryBuffer` in your model and setting it as the buffer of an `gtk::Entry`.

**Reply by gtamas** [reply]
Thanks for all the info and the detailed explanation! Cool, everything works now.

**Reply by chrsoo** [reply]
> Since 0.6, Relm4 has bindings which might be even better for your use-case: https://docs.rs/relm4/latest/relm4/binding/index.html

Would you have an example on how to use this feature to bind text from a model to a label or the name of an icon to a button?

Looking at the `ConnectBinding` documentation, support for this would appear to be available per default but I am unable to figure out on how to use it or find the relevant documentation...

**Reply by chrsoo** [reply]
... and of course I managed to find [an example](https://docs.rs/relm4/latest/src/data_binding/data_binding.rs.html) just after posting. Sorry for the noise!

For posterity the binding is done [like so](https://docs.rs/relm4/latest/src/data_binding/data_binding.rs.html#48) in the view macro:
````rust

gtk::Label::with_binding(&model.text) {
    set_margin_all: 5,
},
````
... with the [model's text property](https://docs.rs/relm4/latest/src/data_binding/data_binding.rs.html#12) required to be of the right binding type:
````rust
struct App {
    ...
    text: StringBinding,
}
````
